samples/compiler_plugins/xnnpack/src/xnnpack/Conversion/StablehloToXnnpack.cpp

```diff
@@ -26,6 +28,58 @@ static llvm::cl::opt<std::string> patternFileName(
     "xnnpack-pattern-file", llvm::cl::desc("file for pattern bytecode"),
     llvm::cl::init(""));
 
+/// Custom constraint invoked from PDL.
+static LogicalResult checkI4RankedTensorType(PatternRewriter &rewriter,
+                                             Value value) {
+  if (auto ty = dyn_cast<RankedTensorType>(value.getType())) {
+    return success(ty.getElementType().isInteger(4));
+  }
+  return failure();
+}
+
+static LogicalResult checkI8RankedTensorType(PatternRewriter &rewriter,
+                                             Value value) {
+  if (auto ty = dyn_cast<RankedTensorType>(value.getType())) {
+    return success(ty.getElementType().isInteger(8));
+  }
+  return failure();
+}
+
+static LogicalResult checkI32RankedTensorType(PatternRewriter &rewriter,
+                                              Value value) {
+  if (auto ty = dyn_cast<RankedTensorType>(value.getType())) {
+    return success(ty.getElementType().isInteger(32));
+  }
+  return failure();
+}
+
+static LogicalResult checkF32RankedTensorType(PatternRewriter &rewriter,
+                                              Value value) {
+  if (auto ty = dyn_cast<RankedTensorType>(value.getType())) {
+    return success(ty.getElementType().isF32());
+  }
+  return failure();
+}
+
+static LogicalResult checkInnermostReduction(PatternRewriter &rewriter,
+                                             Value value) {
+  auto op = value.getDefiningOp<stablehlo::DotGeneralOp>();
```

Comment:
nit: assert or return `notifyMatchFailure` if defining op is not a `DotGeneralOp`